### PR TITLE
Moving WormManager from d2k to Common as ActorSpawnManager

### PR DESCRIPTION
**Pull request by CDVoidwalker**
Works just like old WormManager with addition of being conditional and having Types.

### CODE REVIEW

**Comment by MustaphaTR** [comment]
> Checking for explicit interface violations...
> ActorSpawnManager must explicitly implement the interface member ITick.Tick
> Explicit interface violations: 1
> make: *** [check] Error 1

Also commits should be squashed.

**Comment by GraionDilach** [comment]
>OpenRA.Mods.Common/Traits/ActorSpawner.cs:L25: [SA1513] Statements or elements wrapped in curly brackets must be followed by a blank line.
OpenRA.Mods.Common/Traits/ActorSpawner.cs:L14: [SA1208] System using directives must be placed before all other using directives.
OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs:L21: [SA1001] Invalid spacing around the comma.
OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs:L34: [SA1012] Invalid spacing around the opening curly bracket.
OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs:L34: [SA1013] Invalid spacing around the closing curly bracket.

**Comment by Unit158** [comment]
There's a binary file that shouldn't have been included.

File: mods/d2k/rules/world.yaml [hunk 1]
```diff
@@ -72,7 +72,9 @@ World:
 	BuildingInfluence:
 	ProductionQueueFromSelection:
 		ProductionPaletteWidget: PRODUCTION_PALETTE
-	WormManager:
+	ActorSpawnManager:
+		ActorSignature: sandworm
```
Comment (Unit158) [inline mods/d2k/rules/world.yaml]: Why not just Actor:

File: mods/d2k/rules/world.yaml [hunk 2]
```diff
@@ -72,7 +72,9 @@ World:
...
+		ActorSignNotification: WormSign
```
Comment (Unit158) [inline mods/d2k/rules/world.yaml]: ActorSpawnNotification:?

File: OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+		[Desc("Time (in ticks) between actor spawn.")]
+		public readonly int SpawnInterval = 6000;
+
+		[FieldLoader.Require]
+		[Desc("Name of the actor that will be spawned.")]
+		public readonly string[] ActorSignature = { };
+
+		public readonly string ActorSignNotification = "";
+		public readonly string ActorOwnerPlayer = "Creeps";
```
Comment (Unit158) [inline OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs]: I'd argue that this would be just as clear as "ActorOwner:"

File: OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+		[Desc("Type of ActorSpawner which it connects with")]
```
Comment (Unit158) [inline OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs]: "which it connects with" -> "with which it connects"

File: OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+	}
+
+	public class ActorSpawnManager : ConditionalTrait<ActorSpawnManagerInfo>, ITick, INotifyCreated
+	{
+		readonly ActorSpawnManagerInfo info;
+		readonly Lazy<TraitPair<ActorSpawner>[]> spawnPointActors;
+
+		bool enabled;
+		int spawnCountdown;
+		int actorPresent;
```
Comment (Unit158) [inline OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs]: actorsPresent

File: OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+			if (--spawnCountdown > 0 && actorPresent >= info.Minimum)
+				return;
+
+			spawnCountdown = info.SpawnInterval;
+
+			var actorLocations = new List<WPos>();
+
+			do
+			{
+				// Always spawn at least one actor, plus however many
```
Comment (Unit158) [inline OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs]: "Always spawn at least one actor, plus however many needed to reach the minimum"

File: OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+			if (IsTraitDisabled)
+				return;
+
+			if (!enabled)
+				return;
+
+			if (!spawnPointActors.Value.Any())
+				return;
+
+			// Apparently someone doesn't want actors or the maximum number of actors has been reached
```
Comment (Unit158) [inline OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs]: I'd argue the next line doesn't warrant a comment.

File: OpenRA.Mods.Common/Traits/ActorSpawner.cs [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+using System;
+using System.Collections.Generic;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("An actor with this trait indicates a valid spawn point for actors.")]
+	public class ActorSpawnerInfo : ConditionalTraitInfo
+	{
+		[Desc("Type of ActorSpawner which it connects with")]
```
Comment (Unit158) [inline OpenRA.Mods.Common/Traits/ActorSpawner.cs]: "which it connects with" -> "with which it connects"

File: OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+		[Desc("Maximum number of actors")]
+		public readonly int Maximum = 4;
+
+		[Desc("Time (in ticks) between actor spawn.")]
+		public readonly int SpawnInterval = 6000;
+
+		[FieldLoader.Require]
+		[Desc("Name of the actor that will be spawned.")]
+		public readonly string[] ActorSignature = { };
```
Comment (Unit158) [inline OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs]: ActorSignature -> Actors

File: OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+		public readonly string ActorSignNotification = "";
```
Comment (Unit158) [inline OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs]: SpawnNotification

**Comment by Unit158** [comment]
I haven't tested yet, but the code looks OK, minus some style/naming/language improvements.

File: OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+				new OwnerInit(w.Players.First(x => x.PlayerName == info.ActorOwnerPlayer)),
+				new LocationInit(spawnPoint.Location)
+			}));
+
+			actorPresent++;
+
+			return spawnPoint.CenterPosition;
+		}
+
+		Actor GetRandomSpawnPoint(Actor self)
```
Comment (Unit158) [inline OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs]: I'd argue this should only pass the shared RNG.

File: OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs (hunk 10 above)
Comment (penev92) [reply OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs]: No, that was actually `public readonly string WormSignNotification = "WormSign";`.
So this one probably needs some more work.

Comment (CDVoidwalker) [reply OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs]: [FieldLoader.Require] ? or null?

File: OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs (hunk 3 above)
Comment (CDVoidwalker) [reply OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs]: That was directly copied from WormManager, as with all the comments

File: OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+			if (!enabled)
+				return;
+
+			if (!spawnPointActors.Value.Any())
+				return;
+
+			if (info.Maximum < 1 || actorsPresent >= info.Maximum)
+				return;
+
+			if (--spawnCountdown > 0 && actorsPresent >= info.Minimum)
```
Comment (MunWolf) [inline OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs]: This might not ever happen but there could be an "underflow" here if the second condition never fires.

**Comment by MunWolf** [comment]
This still needs an upgrade rule for the name change.

File: OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs (hunk 12 above)
Comment (CDVoidwalker) [reply OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs]: There's no real case when it would happen, as adding actors is after this check

**Comment by pchote** [comment]
Can you please also squash the fixup commits into the main one?

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj [hunk 13]
```diff
@@ -837,22 +839,14 @@
   </ItemGroup>
   <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
   <Target Name="AfterBuild">
-    <MakeDir Directories="$(SolutionDir)mods/common/"/>
```
Comment (pchote) [inline OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: Please unstage these unrelated project file formatting changes.

File: OpenRA.Mods.Common/Traits/ActorSpawner.cs [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("An actor with this trait indicates a valid spawn point for actors.")]
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/ActorSpawner.cs]: Can you please have this mention that it is specifically for `ActorSpawnManager`?

File: OpenRA.Mods.Common/Traits/ActorSpawner.cs [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+		[Desc("Type of ActorSpawner with which it connects.")]
+		public readonly HashSet<string> Types = new HashSet<string>() { "" };
+
+		public override object Create(ActorInitializer init) { return new ActorSpawner(this); }
+	}
+
+	public class ActorSpawner : ConditionalTrait<ActorSpawnerInfo>
+	{
+		public ActorSpawner(ActorSpawnerInfo info) : base(info)
+		{ }
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/ActorSpawner.cs]: Our code style conventions for cases like this is
```csharp
public ActorSpawner(ActorSpawnerInfo info)
	: base(info) { }
```

File: OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Controls the spawning of actors. Attach this to the world actor.")]
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs]: Can you please change this to something like "Controls spawning of map creep actors"?  There are other types of actor spawning in the game, so this could otherwise be confusing.

File: OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+		public readonly int Maximum = 4;
+
+		[Desc("Time (in ticks) between actor spawn.")]
+		public readonly int SpawnInterval = 6000;
+
+		[FieldLoader.Require]
+		[Desc("Name of the actor that will be spawned.")]
+		public readonly string[] Actors = { };
+
+		public readonly string ActorSignNotification = "";
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs]: This isn't used, so please remove it.

File: OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+
+		[Desc("Maximum number of actors")]
+		public readonly int Maximum = 4;
+
+		[Desc("Time (in ticks) between actor spawn.")]
+		public readonly int SpawnInterval = 6000;
+
+		[FieldLoader.Require]
+		[Desc("Name of the actor that will be spawned.")]
+		public readonly string[] Actors = { };
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs]: These are actor references, so please attach the `[ActorReference]` annotation.

File: OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+		public readonly int Minimum = 0;
+
+		[Desc("Maximum number of actors")]
+		public readonly int Maximum = 4;
+
+		[Desc("Time (in ticks) between actor spawn.")]
+		public readonly int SpawnInterval = 6000;
+
+		[FieldLoader.Require]
+		[Desc("Name of the actor that will be spawned.")]
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs]: Please mention that the spawned actor types are selected randomly from this list - otherwise it could be interpreted as spawning them all at once.

File: OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+
+			if (info.Maximum < 1 || actorsPresent >= info.Maximum)
+				return;
+
+			if (--spawnCountdown > 0 && actorsPresent >= info.Minimum)
+				return;
+
+			spawnCountdown = info.SpawnInterval;
+
+			var actorLocations = new List<WPos>();
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs]: This isn't used for anything?

File: OpenRA.Mods.D2k/OpenRA.Mods.D2k.csproj [hunk 21]
```diff
@@ -121,22 +119,14 @@
   </ItemGroup>
   <Import Project="$(MSBuildBinPath)\Microsoft.CSharp.targets" />
   <Target Name="AfterBuild">
-    <MakeDir Directories="$(SolutionDir)mods/d2k/"/>
```
Comment (pchote) [inline OpenRA.Mods.D2k/OpenRA.Mods.D2k.csproj]: Please unstage unrelated project file changes.

**Review by pchote — COMMENTED** [review]
I haven't tested functionality, but I have left a few style nits.
As @MunWolf mentioned this will also need an upgrade rule to manage the rename.

File: OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs (hunk 16 above)
Comment (CDVoidwalker) [reply OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs]: Yeah, but also those actors aren't creep only, depends on what is set in Owner of the manager, so "Controls spawning of the map actors"?

**Comment by pchote** [comment]
It would be good to keep this moving and to get it merged before it loses momentum.

Do you need any help with this?

File: OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs (hunk 16 above)
Comment (pchote) [reply OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs]: I would go with "Controls the spawning of specified actor types" to keep the distinction vs e.g. `SpawnMpUnits` clear.

**Comment by CDVoidwalker** [comment]
Will try to get it done for tomorrow

**Comment by GraionDilach** [comment]
>OpenRA.Mods.Common/Traits/ActorSpawner.cs:L29: [SP2000] Invalid spacing at the end of the line.

**Comment by CDVoidwalker** [comment]
needs review

**Comment by MustaphaTR** [comment]
Needs rebase.

**Review by GraionDilach — APPROVED** [review]
:+1: but someone will need to update the date of the upgrade rule.

**Comment by abcdefg30** [comment]
Also needs a rebase yet again. :/

**Comment by pchote** [comment]
I want to see this merged for the playtest, so I have amended the branch with a few changes:
* Rebased and updated upgrade rule date.
* Replaced the default "" type with the same "overlaps or types is empty" check that we use everywhere else.
* Replaced the `Lazy<TraitPair<ActorSpawner>[]>` with a more modern Created lookup.

**Review by pchote — DISMISSED** [review]
LGTM and works as advertised ingame.

My amendments will need a :+1: before merging.

File: OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+using System.Linq;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Controls the spawning of specified actor types. Attach this to the world actor.")]
+	public class ActorSpawnManagerInfo : ConditionalTraitInfo, Requires<MapCreepsInfo>
+	{
+		[Desc("Minimum number of actors")]
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs]: Missing fullstop.

File: OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+		public readonly int Minimum = 0;
+
+		[Desc("Maximum number of actors")]
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs]: Missing fullstop.

File: OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs [hunk 24]
```diff
@@ -0,0 +1,116 @@
...
+			}));
+
+			actorsPresent++;
+
+			return spawnPoint.CenterPosition;
+		}
+
+		Actor GetRandomSpawnPoint(Support.MersenneTwister random)
+		{
+			return spawnPointActors.Random(random).Actor;
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs]: ActorSpawner traits are also conditional, but this is ignored in this current revision. IMO it'd make the most sense to filter the spawnpoints here.

**Comment by GraionDilach** [comment]
I'm actually not certain how much sense it makes to cache the suitable actor spawners on creation. Not doing that would allow one to enable map creeps midgame based on some initial condition.

**Comment by pchote** [comment]
Yes, that makes sense.  Certain mods or minigames might want to use player-built actors as spawn points.

**Comment by Arular101** [comment]
Please don't forget to update the copyright notice year.

**Comment by reaperrr** [comment]
Adressed all remaining style nits, replaced caching-at-creation with evaluation right before spawn attempt, rebased.

**Comment by abcdefg30** [comment]
> OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs(48,29): error CS0169: Warning as Error: The private field `OpenRA.Mods.Common.Traits.ActorSpawnManager.spawnPointActors' is never used

**Comment by reaperrr** [comment]
Fixed.

File: OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs [hunk 25]
```diff
@@ -0,0 +1,112 @@
...
+
+			if (info.Maximum < 1 || actorsPresent >= info.Maximum)
+				return;
+
+			if (--spawnCountdown > 0 && actorsPresent >= info.Minimum)
+				return;
+
+			var spawnPoint = GetRandomSpawnPoint(self.World, self.World.SharedRandom);
+
+			if (!enabled)
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs]: Does caching the nullcheck to a boolean for this usage worth it?

**Comment by reaperrr** [comment]
Updated.

> Does caching the nullcheck to a boolean for this usage worth it?

No, but I also forgot to keep the MapCreeps.Enabled check, so now I'm using `enabled` to cache that in the ctor (ASM Requires MapCreeps, so the latter should be constructed before the former) but use `spawnPoint == null` to abort later if there's no enabled spawnpoint.

File: OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs (hunk 12 above)
Comment (MunWolf) [reply OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs]: I meant that --spawnCountdown might eventually underflow to a really big value after -2147483648 ticks... which will probably never happen since it will take way too long.

Comment (reaperrr) [reply OpenRA.Mods.Common/Traits/World/ActorSpawnManager.cs]: @MunWolf Unless I'm miscalculating, we're talking about nearly 12000 hours at maximum game speed (50 ticks/sec).

I dare say whoever manages to set things up weirdly enough to trigger this, deserves the crash ;)

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
I've proof-read and tested my own updates. If it turns out I overlooked anything, I'll take responsibility and file a follow-up.

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs [hunk 26]
```diff
@@ -1635,6 +1635,21 @@ internal static void UpgradeActorRules(ModData modData, int engineVersion, ref L
 					}
 				}
 
+				if (engineVersion < 20180225)
```
Comment (GraionDilach) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: Needs an update on this due to new release (?).

Comment (reaperrr) [reply OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: Yeah, but it's not the only case, I think. I'll fix it either way.

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/3a17c2590602a423b33f9e6ebbb9e36d62dcb982...ad98895675fe2bee9a7000cd607b74bffe669928
